likelihood.py

```diff
@@ -0,0 +1,232 @@
+        Delta_r = log(rmax/self.rmin) / (self.Nmax - 1)
+        i_arr = np.arange(self.Nmax)
+        rtab = self.rmin * exp(Delta_r * i_arr)
+
+        self.kbins3 = self.k0 * exp(Delta * i_arr)
+        self.tmp_factor = exp(-1.*b*i_arr*Delta)
+        self.tmp_factor2 = exp(-1.*bR*i_arr*Delta_r)[:,np.newaxis]
+
+        jsNm = np.arange(-self.Nmax//2,self.Nmax//2+1,1)
+        self.etam = b + 2*1j*pi*(jsNm)/self.Nmax/Delta
+
+        def J_func(r,nu):
+            gam = special.gamma(2+nu)
+            r_pow = r**(-3.-1.*nu)
+            sin_nu = np.sin(pi*nu/2.)
+            J0 = -1.*sin_nu*r_pow*gam/(2.*pi**2.)
+            return J0
+
+        j0 = J_func(rtab.reshape(-1,1),self.etam.reshape(1,-1))
+        self.J0_arr = j0[:,:,np.newaxis]
+
+        self.etamR = bR + 2*1j*pi*(jsNm)/self.Nmax/Delta_r
+
+        def Jk_func(k,nu):
+            gam = special.gamma(2+nu)
+            k_pow = k**(-3.-1.*nu)
+            sin_nu = np.sin(pi*nu/2.)
+            J0k = -1.*k_pow*gam*sin_nu*(4.*pi)
+            return J0k
+
+        j0k = Jk_func(self.kbins3.reshape(-1,1),self.etamR.reshape(1,-1))
+        self.J0k_arr = j0k[:,:,np.newaxis]
+
+        #Precompute inverse covariance matrix
+        self.invcov = np.linalg.inv(self.cov)
+
+        # Compute window response matrix
+        self.response_matrix = np.zeros((self.ksize,self.Nmax))
+        for i in range(self.Nmax):
+            tmp_resp0 = self.window_response(i)
+            self.response_matrix[:,i] = tmp_resp0[0]
+
+        # Now convolve with window function
+        self.invcovW = np.matmul(self.response_matrix.T,self.invcov)
+        self.invcovWW = np.matmul(self.response_matrix.T,np.matmul(self.invcov,self.response_matrix))
+
+    def window_response(self,k_index):
+        """Window function. This is from 1607.03150 and includes only the monopole.
+        Accurate for k > 0.0015."""
+        Nmax = self.Nmax
+        k0 = self.k0
+
+        Pdiscrin0 = np.zeros(Nmax)
+        Pdiscrin0[k_index] = 1
+
+        cm0 = np.fft.fft(Pdiscrin0)/ Nmax
+        cmsym0 = np.zeros(Nmax+1,dtype=np.complex_)
+
+        all_i = np.arange(Nmax+1)
+        f = (all_i+2-Nmax//2) < 1
+        cmsym0[f] = k0**(-self.etam[f])*np.conjugate(cm0[-all_i[f]+Nmax//2])
+        cmsym0[~f] = k0**(-self.etam[~f])*cm0[all_i[~f]-Nmax//2]
+
+        cmsym0[-1] = cmsym0[-1] / 2
+        cmsym0[0] = cmsym0[0] / 2
+
+        xi0 = np.real(cmsym0*self.J0_arr).sum(axis=1)
+        Xidiscrin0 = (xi0*self.W0)*self.tmp_factor2
+
+        cmr0 = np.fft.fft(Xidiscrin0)/ Nmax
+
+        cmsymr0 = np.zeros(Nmax+1,dtype=np.complex_)
+
+        arr_i = np.arange(Nmax+1)
+        f = (arr_i+2-Nmax//2)<1
+
+        cmsymr0[f] = self.rmin**(-self.etamR[f])*np.conjugate(cmr0[-arr_i[f] + Nmax//2])
+        cmsymr0[~f] = self.rmin**(-self.etamR[~f])* cmr0[arr_i[~f] - Nmax//2]
+
+        cmsymr0[-1] = cmsymr0[-1] / 2
+        cmsymr0[0] = cmsymr0[0] / 2
+
+        P0t = np.real(cmsymr0*self.J0k_arr).sum(axis=1)
```

Comment:
`cmsymr0` is (129, ), needs an additional dim to be (129, 1) and
multiply with `J0k_arr` (128, 129, 1).

`P0t` will be (128, 1)

```python
        P0t = np.real(cmsymr0[:, None]*self.J0k_arr).sum(axis=1)
```